File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/java/io/github/lishangbu/avalon/dataset/controller/TypeDamageRelationControllerTest.java
package io.github.lishangbu.avalon.dataset.controller;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation;
import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation.TypeDamageRelationId;
import io.github.lishangbu.avalon.dataset.service.TypeDamageRelationService;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class TypeDamageRelationControllerTest {
    @Mock private TypeDamageRelationService typeDamageRelationService;
    @InjectMocks private TypeDamageRelationController typeDamageRelationController;

    @Test
    void getTypeDamageRelationPage_delegatesToService() {
        Pageable pageable = PageRequest.of(0, 5);
        Page<TypeDamageRelation> page = new PageImpl<>(List.of(new TypeDamageRelation()));
        when(typeDamageRelationService.getPageByCondition(
                        any(TypeDamageRelation.class), eq(pageable)))
                .thenReturn(page);

        Page<TypeDamageRelation> result =
                typeDamageRelationController.getTypeDamageRelationPage(pageable, 1L, 2L, 2.0f);

        assertSame(page, result);
        verify(typeDamageRelationService)
                .getPageByCondition(any(TypeDamageRelation.class), eq(pageable));
    }

    @Test
    void listTypeDamageRelations_delegatesToService() {
        List<TypeDamageRelation> list = List.of(new TypeDamageRelation());
        when(typeDamageRelationService.listByCondition(any(TypeDamageRelation.class)))
                .thenReturn(list);

        List<TypeDamageRelation> result =
                typeDamageRelationController.listTypeDamageRelations(1L, 2L, 0.5f);

        assertSame(list, result);
        verify(typeDamageRelationService).listByCondition(any(TypeDamageRelation.class));
    }

    @Test
    void save_delegatesToService() {
        TypeDamageRelation relation = new TypeDamageRelation();
        when(typeDamageRelationService.save(relation)).thenReturn(relation);

        TypeDamageRelation result = typeDamageRelationController.save(relation);

        assertSame(relation, result);
        verify(typeDamageRelationService).save(relation);
    }

    @Test
    void update_delegatesToService() {
        TypeDamageRelation relation = new TypeDamageRelation();
        when(typeDamageRelationService.update(relation)).thenReturn(relation);

        TypeDamageRelation result = typeDamageRelationController.update(relation);

        assertSame(relation, result);
        verify(typeDamageRelationService).update(relation);
    }

    @Test
    void deleteById_delegatesToService() {
        typeDamageRelationController.deleteById(1L, 2L);
        verify(typeDamageRelationService).removeById(any(TypeDamageRelationId.class));
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/kotlin/io/github/lishangbu/avalon/dataset/controller/TypeDamageRelationControllerTest.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelation
import io.github.lishangbu.avalon.dataset.entity.TypeDamageRelationId
import io.github.lishangbu.avalon.dataset.service.TypeDamageRelationService
import org.babyfish.jimmer.Page
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable

class TypeDamageRelationControllerTest {
    @Test
    fun getTypeDamageRelationPage_delegatesToService() {
        val service = FakeTypeDamageRelationService()
        val controller = TypeDamageRelationController(service)
        val pageIndex = 0
        val pageSize = 5
        val pageable = PageRequest.of(pageIndex, pageSize)
        val page: Page<TypeDamageRelation> = Page(listOf(TypeDamageRelation()), 1, 1)
        service.pageResult = page

        val result = controller.getTypeDamageRelationPage(pageable, 1L, 2L, 2.0f)

        assertSame(page, result)
        assertSame(pageable, service.pageable)
        assertEquals(1L, service.pageAttackingTypeId)
        assertEquals(2L, service.pageDefendingTypeId)
        assertEquals(2.0f, service.pageMultiplier)
    }

    @Test
    fun listTypeDamageRelations_delegatesToService() {
        val service = FakeTypeDamageRelationService()
        val controller = TypeDamageRelationController(service)
        val list = listOf(TypeDamageRelation())
        service.listResult = list

        val result = controller.listTypeDamageRelations(1L, 2L, 0.5f)

        assertSame(list, result)
        assertEquals(1L, service.listAttackingTypeId)
        assertEquals(2L, service.listDefendingTypeId)
        assertEquals(0.5f, service.listMultiplier)
    }

    @Test
    fun save_delegatesToService() {
        val service = FakeTypeDamageRelationService()
        val controller = TypeDamageRelationController(service)
        val relation = TypeDamageRelation()
        service.saveResult = relation

        val result = controller.save(relation)

        assertSame(relation, result)
        assertSame(relation, service.savedRelation)
    }

    @Test
    fun update_delegatesToService() {
        val service = FakeTypeDamageRelationService()
        val controller = TypeDamageRelationController(service)
        val relation = TypeDamageRelation()
        service.updateResult = relation

        val result = controller.update(relation)

        assertSame(relation, result)
        assertSame(relation, service.updatedRelation)
    }

    @Test
    fun deleteById_delegatesToService() {
        val service = FakeTypeDamageRelationService()
        val controller = TypeDamageRelationController(service)

        controller.deleteById(1L, 2L)

        assertNotNull(service.removedId)
        assertEquals(1L, service.removedId!!.attackingTypeId)
        assertEquals(2L, service.removedId!!.defendingTypeId)
    }

    private class FakeTypeDamageRelationService : TypeDamageRelationService {
        var pageAttackingTypeId: Long? = null
        var pageDefendingTypeId: Long? = null
        var pageMultiplier: Float? = null
        var pageable: Pageable? = null

        var listAttackingTypeId: Long? = null
        var listDefendingTypeId: Long? = null
        var listMultiplier: Float? = null

        var savedRelation: TypeDamageRelation? = null
        var updatedRelation: TypeDamageRelation? = null
        var removedId: TypeDamageRelationId? = null

        var pageResult: Page<TypeDamageRelation> = Page(emptyList(), 0, 0)
        var listResult: List<TypeDamageRelation> = emptyList()
        var saveResult: TypeDamageRelation = TypeDamageRelation()
        var updateResult: TypeDamageRelation = TypeDamageRelation()

        override fun getPageByCondition(
            attackingTypeId: Long?,
            defendingTypeId: Long?,
            multiplier: Float?,
            pageable: Pageable,
        ): Page<TypeDamageRelation> {
            pageAttackingTypeId = attackingTypeId
            pageDefendingTypeId = defendingTypeId
            pageMultiplier = multiplier
            this.pageable = pageable
            return pageResult
        }

        override fun save(relation: TypeDamageRelation): TypeDamageRelation {
            savedRelation = relation
            return saveResult
        }

        override fun update(relation: TypeDamageRelation): TypeDamageRelation {
            updatedRelation = relation
            return updateResult
        }

        override fun removeById(id: TypeDamageRelationId) {
            removedId = id
        }

        override fun listByCondition(
            attackingTypeId: Long?,
            defendingTypeId: Long?,
            multiplier: Float?,
        ): List<TypeDamageRelation> {
            listAttackingTypeId = attackingTypeId
            listDefendingTypeId = defendingTypeId
            listMultiplier = multiplier
            return listResult
        }
    }
}
